File: server/internal/handlers/inbox_context_test.go
package handlers_test

import (
	"strings"
	"testing"
	"time"

	"github.com/cocofhu/approving/internal/models"
)

func TestRunInboxContextGate(t *testing.T) {
	h := newHarness(t)
	now := time.Now()
	h.db.Create(&models.Run{ID: "ic-gate", Status: "waiting_human", StartedAt: now, Graph: models.Graph{
		Nodes: []models.Node{
			{ID: "visual", Type: "visual", Config: map[string]any{}},
			{ID: "gate", Type: "human_gate", Config: map[string]any{
				"body_template": "preview {{nodes.visual.outputs.page}}",
			}},
		},
	}})
	h.db.Create(&models.Gate{RunID: "ic-gate", NodeID: "gate", Iteration: 1, Resolved: false, RequestedAt: now})
	h.db.Create(&models.StateRun{
		RunID: "ic-gate", NodeID: "visual", Iteration: 1, Status: "completed",
		Outputs: map[string]any{"page": "<html>v1</html>"},
		Events:  []models.AcpEvent{{Kind: "message", Text: "should-not-appear"}},
	})
	h.db.Create(&models.Artifact{
		ID: "ic-art", RunID: "ic-gate", NodeID: "visual", Name: "page.html", Kind: "html",
		Content: strings.Repeat("x", 5000), SizeBytes: 5000,
	})

	w := h.do("GET", "/api/runs/ic-gate/inbox-context?nodeId=gate&iteration=1", nil)
	if w.Code != 200 {
		t.Fatalf("gate inbox-context: %d %s", w.Code, w.Body)
	}
	body := w.Body.String()
	for _, want := range []string{`"type":"gate"`, `"nodes"`, `"artifacts"`, `"nodeExecutions"`, "page.html", `\u003chtml\u003ev1\u003c/html\u003e`} {
		if !strings.Contains(body, want) {
			t.Errorf("gate response missing %q in %s", want, body)
		}
	}
	if strings.Contains(body, strings.Repeat("x", 100)) {
		t.Error("gate inbox-context must not inline artifact content")
	}
	if strings.Contains(body, "should-not-appear") {
		t.Error("nodeExecutions must be slim (no events)")
	}
}

func TestRunInboxContextClarify(t *testing.T) {
	h := newHarness(t)
	now := time.Now()
	h.db.Create(&models.Run{
		ID: "ic-clarify", Status: "waiting_human", StartedAt: now,
		Graph: models.Graph{Nodes: []models.Node{
			{ID: "research", Type: "research", Label: "调研"},
			{ID: "react", Type: "react", Label: "需求澄清", Config: map[string]any{
				"prompt": "upstream {{nodes.research.outputs.research}}",
			}},
		}},
	})
	h.db.Create(&models.ReactConversation{
		RunID: "ic-clarify", NodeID: "react", Iteration: 1, Done: false,
		Messages: []models.ReactMessage{{Role: "agent", Text: "question?", At: now.Format(time.RFC3339)}},
	})
	h.db.Create(&models.StateRun{RunID: "ic-clarify", NodeID: "react", Iteration: 1, Status: "waiting_human"})
	h.db.Create(&models.StateRun{
		RunID: "ic-clarify", NodeID: "research", Iteration: 1, Status: "completed",
		Outputs: map[string]any{"research": `{"summary":"ok"}`},
		Events:  []models.AcpEvent{{Kind: "message", Text: "should-not-appear"}},
	})
	h.db.Create(&models.Artifact{
		ID: "ic-clarify-art", RunID: "ic-clarify", NodeID: "research", Name: "research.json", Kind: "json",
		Content: strings.Repeat("x", 5000), SizeBytes: 5000,
	})

	w := h.do("GET", "/api/runs/ic-clarify/inbox-context?nodeId=react&iteration=1", nil)
	if w.Code != 200 {
		t.Fatalf("clarify inbox-context: %d %s", w.Code, w.Body)
	}
	body := w.Body.String()
	for _, want := range []string{
		`"type":"clarify"`, `"status"`, `"clarify"`, "question?", "需求澄清",
		`"nodes"`, `"artifacts"`, `"nodeExecutions"`, "research.json",
	} {
		if !strings.Contains(body, want) {
			t.Errorf("clarify response missing %q in %s", want, body)
		}
	}
	if strings.Contains(body, strings.Repeat("x", 100)) {
		t.Error("clarify inbox-context must not inline artifact content")
	}
	if strings.Contains(body, "should-not-appear") {
		t.Error("nodeExecutions must be slim (no events)")
	}
	// Idle (no in-memory session): reactSessions omitted — presence when busy is
	// covered by TestReactSessionsDTO + GatesInbox hard-load restore vitest.
	if strings.Contains(body, `"reactSessions"`) {
		t.Error("idle clarify inbox-context should omit empty reactSessions")
	}
}

func TestRunInboxContextClarifyResearchReview(t *testing.T) {
	h := newHarness(t)
	now := time.Now()
	h.db.Create(&models.Run{
		ID: "ic-research", Status: "waiting_human", StartedAt: now,
		Graph: models.Graph{Nodes: []models.Node{
			{ID: "research_1", Type: "research", Label: "调研结论"},
		}},
	})
	h.db.Create(&models.ReactConversation{
		RunID: "ic-research", NodeID: "research_1", Iteration: 1, Done: false,
		Messages: []models.ReactMessage{{Role: "agent", Text: "请复审调研", At: now.Format(time.RFC3339)}},
	})
	h.db.Create(&models.StateRun{
		RunID: "ic-research", NodeID: "research_1", Iteration: 1, Status: "waiting_human",
		Outputs: map[string]any{"research": `{"summary":"root cause"}`},
	})
	h.db.Create(&models.Artifact{
		ID: "ic-res-art", RunID: "ic-research", NodeID: "research_1", Name: "research.json", Kind: "json",
		Content: `{"summary":"root cause"}`, SizeBytes: 24,
	})

	w := h.do("GET", "/api/runs/ic-research/inbox-context?nodeId=research_1&iteration=1", nil)
	if w.Code != 200 {
		t.Fatalf("research clarify inbox-context: %d %s", w.Code, w.Body)
	}
	body := w.Body.String()
	for _, want := range []string{
		`"type":"clarify"`, `"nodes"`, `"artifacts"`, `"nodeExecutions"`,
		"research_1", "research.json", "root cause",
	} {
		if !strings.Contains(body, want) {
			t.Errorf("research clarify response missing %q in %s", want, body)
		}
	}
}

func TestRunInboxContextNotFound(t *testing.T) {
	h := newHarness(t)
	h.db.Create(&models.Run{ID: "ic-none", Status: "waiting_human", StartedAt: time.Now(), Graph: models.Graph{}})

	if w := h.do("GET", "/api/runs/ic-none/inbox-context?nodeId=x&iteration=1", nil); w.Code != 404 {
		t.Fatalf("no pending: %d", w.Code)
	}
	if w := h.do("GET", "/api/runs/ic-none/inbox-context?nodeId=x", nil); w.Code != 400 {
		t.Fatalf("missing iteration: %d", w.Code)
	}
	if w := h.do("GET", "/api/runs/ic-none/inbox-context?iteration=1", nil); w.Code != 400 {
		t.Fatalf("missing nodeId: %d", w.Code)
	}
}

func TestRunInboxContextIterationMismatch(t *testing.T) {
	h := newHarness(t)
	now := time.Now()
	h.db.Create(&models.Run{ID: "ic-iter", Status: "waiting_human", StartedAt: now, Graph: models.Graph{
		Nodes: []models.Node{{ID: "gate", Type: "human_gate"}},
	}})
	h.db.Create(&models.Gate{RunID: "ic-iter", NodeID: "gate", Iteration: 2, Resolved: false, RequestedAt: now})

	if w := h.do("GET", "/api/runs/ic-iter/inbox-context?nodeId=gate&iteration=1", nil); w.Code != 404 {
		t.Fatalf("iteration mismatch should 404: %d", w.Code)
	}
}

File: web/src/lib/inbox/inboxContext.test.ts
import { describe, expect, it } from 'vitest'
import { adaptInboxContextToRun } from './inboxContext'

describe('adaptInboxContextToRun', () => {
  it('maps gate context to Run subset', () => {
    const run = adaptInboxContextToRun(
      {
        type: 'gate',
        nodes: [{ id: 'gate', type: 'human_gate', label: 'G', position: { x: 0, y: 0 }, config: {} }],
        artifacts: [{ id: 'a1', name: 'plan.json', kind: 'json', nodeId: 'plan', runId: 'r1', workflowName: 'wf', sizeBytes: 1, createdAt: '' }],
        nodeExecutions: { visual: [{ nodeId: 'visual', status: 'completed', iteration: 1, outputs: { page: '<p/>' } }] },
      },
      'r1',
    )
    expect(run.id).toBe('r1')
    expect(run.nodes).toHaveLength(1)
    expect(run.artifacts[0].name).toBe('plan.json')
    expect(run.nodeExecutions?.visual[0].outputs?.page).toBe('<p/>')
  })

  it('maps clarify context to clarifyByNode + status', () => {
    const run = adaptInboxContextToRun(
      {
        type: 'clarify',
        status: 'waiting_human',
        clarify: {
          nodeId: 'react',
          iteration: 2,
          turns: [{ role: 'agent', text: 'hi', at: '2026-01-01T00:00:00Z' }],
          done: false,
          label: '澄清',
          previewArtifact: 'page.html',
        },
      },
      'r2',
    )
    expect(run.status).toBe('waiting_human')
    expect(run.clarifyByNode?.react.turns).toHaveLength(1)
    expect(run.clarifyByNode?.react.previewArtifact).toBe('page.html')
    expect(run.artifacts).toEqual([])
  })

  it('maps clarify product fields so StructuredProductPanel can mount', () => {
    const run = adaptInboxContextToRun(
      {
        type: 'clarify',
        status: 'waiting_human',
        nodes: [
          { id: 'research_1', type: 'research', label: '调研', position: { x: 0, y: 0 }, config: {} },
          { id: 'react', type: 'react', label: '澄清', position: { x: 1, y: 0 }, config: {} },
        ],
        artifacts: [
          {
            id: 'a1',
            name: 'research.json',
            kind: 'json',
            nodeId: 'research_1',
            runId: 'r3',
            workflowName: 'wf',
            sizeBytes: 12,
            createdAt: '',
          },
        ],
        nodeExecutions: {
          research_1: [
            {
              nodeId: 'research_1',
              status: 'waiting_human',
              iteration: 1,
              outputs: { research: '{"summary":"ok"}' },
            },
          ],
        },
        clarify: {
          nodeId: 'research_1',
          iteration: 1,
          turns: [{ role: 'agent', text: '请复审', at: '2026-01-01T00:00:00Z' }],
          done: false,
          label: '调研结论',
        },
      },
      'r3',
    )
    expect(run.nodes?.find((n) => n.id === 'research_1')?.type).toBe('research')
    expect(run.artifacts[0].name).toBe('research.json')
    expect(run.nodeExecutions?.research_1[0].outputs?.research).toContain('ok')
    expect(run.clarifyByNode?.research_1.done).toBe(false)
  })

  it('maps clarify reactSessions for refresh-resume', () => {
    const run = adaptInboxContextToRun(
      {
        type: 'clarify',
        status: 'waiting_human',
        reactSessions: {
          react: {
            kind: 'clarify',
            waiting: 1,
            busy: true,
            items: [{ id: 'q2', text: '乙' }],
            activeItem: { id: 'q1', text: '甲' },
          },
        },
        clarify: {
          nodeId: 'react',
          iteration: 1,
          turns: [],
          done: false,
          label: '澄清',
        },
      },
      'r-sess',
    )
    expect(run.reactSessions?.react?.busy).toBe(true)
    expect(run.reactSessions?.react?.items?.[0]?.id).toBe('q2')
    expect(run.reactSessions?.react?.activeItem?.text).toBe('甲')
  })

  it('normalizes null clarify turns to an empty array', () => {
    const run = adaptInboxContextToRun(
      {
        type: 'clarify',
        status: 'waiting_human',
        nodes: [{ id: 'predev', type: 'approve', label: 'Approve', position: { x: 0, y: 0 }, config: {} }],
        artifacts: [],
        nodeExecutions: {},
        clarify: {
          nodeId: 'predev',
          iteration: 1,
          turns: null as unknown as [],
          done: false,
          label: 'Approve',
        },
      },
      'r-empty',
    )
    expect(run.clarifyByNode?.predev.turns).toEqual([])
  })
})
